pkg/store/bucket.go

```diff
@@ -787,29 +787,77 @@ func (b *bucketBlock) Close() error {
 }
 
 type bucketIndexReader struct {
-	ctx   context.Context
-	block *bucketBlock
-	dec   *index.DecoderV1
+	logger log.Logger
+	ctx    context.Context
+	block  *bucketBlock
+	dec    *index.DecoderV1
 
 	loadedPostings []*lazyPostings
 	loadedSeries   map[uint64][]byte
 }
 
-func newBucketIndexReader(ctx context.Context, block *bucketBlock) *bucketIndexReader {
+func newBucketIndexReader(ctx context.Context, logger log.Logger, block *bucketBlock) *bucketIndexReader {
 	r := &bucketIndexReader{
-		ctx:   ctx,
-		block: block,
-		dec:   &index.DecoderV1{},
+		logger: logger,
+		ctx:    ctx,
+		block:  block,
+		dec:    &index.DecoderV1{},
 
 		loadedSeries: map[uint64][]byte{},
 	}
 	r.dec.SetSymbolTable(r.block.symbols)
 	return r
 }
 
+func (r *bucketIndexReader) preloadPostings() error {
+	ps := r.loadedPostings
+
+	if len(ps) == 0 {
+		return nil
+	}
+	sort.Slice(ps, func(i, j int) bool {
+		return ps[i].ptr.Start < ps[j].ptr.Start
+	})
+
+	var g run.Group
+
+	const maxPostingsGap = 512 * 1024
+
+	j := 0
+	k := 0
+	for k < len(ps) {
+		j = k
+		k++
+
+		start, end := ps[j].ptr.Start, ps[j].ptr.End
+
+		// Keep growing the range until the end or we encounter a large gap.
+		for ; k < len(ps); k++ {
+			if end+maxPostingsGap < ps[k].ptr.Start {
+				break
+			}
+			end = ps[k].ptr.End
+		}
+		part := ps[j:k]
+
+		ctx, cancel := context.WithCancel(r.ctx)
```

Comment:
This is a change as well. Per discussion with Peter, the intended usage of the group is that each actor has its own cancel function that does not affect other actors. Hence, we've to create one context each.